internal/agent/agent.go

```diff
@@ -100,45 +108,99 @@ func (a *Agent) Run(ctx context.Context) error {
 	defer tpmChannel.Close()
 
 	// create file io writer and reader
-	deviceWriter, _ := initializeFileIO(a.config)
+	deviceWriter, deviceReader := initializeFileIO(a.config)
 
-	// create config controller
-	controller := config.NewController(
+	// create status manager
+	statusManager := status.NewManager(deviceName, tpmChannel, &executer.CommonExecuter{})
+
+	// TODO: this needs tuned
+	backoff := wait.Backoff{
+		Cap:      3 * time.Minute,
+		Duration: 10 * time.Second,
+		Factor:   1.5,
+		Steps:    24,
+	}
+
+	bootstrap := device.NewBootstrap(
+		deviceName,
+		deviceWriter,
+		deviceReader,
+		statusManager,
 		enrollmentClient,
-		a.config.EnrollmentEndpoint,
-		a.config.EnrollmentUIEndpoint,
 		a.config.ManagementEndpoint,
+		a.config.EnrollmentUIEndpoint,
 		a.config.Cacert,
-		a.config.GeneratedCert,
 		a.config.Key,
+		a.config.GeneratedCert,
+		backoff,
+		a.log,
+		a.config.LogPrefix,
+	)
+
+	// bootstrap
+	if err := bootstrap.Initialize(ctx); err != nil {
+		return fmt.Errorf("bootstrap failed: %w", err)
+	}
+
+	// create the management client
+	managementClient, err := newManagementClient(a.config)
+	if err != nil {
+		return err
+	}
+
+	statusManager.SetClient(managementClient)
+
+	// create spec manager
+	specManager := spec.NewManager(
+		deviceName,
+		currentSpecFilePath,
+		desiredSpecFilePath,
+		deviceWriter,
+		deviceReader,
+		managementClient,
+		backoff,
+		a.log,
+		a.config.LogPrefix,
+	)
+
+	// create config controller
+	controller := config.NewController(
 		deviceWriter,
 		csr,
+		a.log,
 		a.config.LogPrefix,
 	)
 
 	// create agent
 	agent := device.NewAgent(
 		deviceName,
+		deviceWriter,
+		statusManager,
+		specManager,
 		a.config.SpecFetchInterval,
 		a.config.StatusUpdateInterval,
-		a.config.Cacert,
-		a.config.GeneratedCert,
-		a.config.Key,
-		a.config.ManagementEndpoint,
-		tpmChannel,
-		&executer.CommonExecuter{},
-		a.config.LogPrefix,
 		controller,
+		a.log,
+		a.config.LogPrefix,
 	)
 
-	go func() {
-		if err := agent.Run(ctx); err != nil {
-			a.log.Fatalf("%s: %v", a.config.LogPrefix, err)
-		}
-	}()
+	return agent.Run(ctx)
+}
 
-	<-ctx.Done()
-	return nil
+func newEnrollmentClient(cfg *Config) (*client.Enrollment, error) {
+	httpClient, err := client.NewWithResponses(cfg.EnrollmentEndpoint, cfg.Cacert, cfg.EnrollmentCertFile, cfg.Key)
```

Comment:
This is going to crash due TLS private key does not match public key.

```suggestion
	httpClient, err := client.NewWithResponses(cfg.EnrollmentEndpoint, cfg.Cacert, cfg.EnrollmentCertFile, cfg.EnrollmentKeyFile)
```

Reply:
good catch